sql/src/main/java/io/crate/metadata/functions/Signature.java

```diff
@@ -0,0 +1,168 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata.functions;
+
+import io.crate.common.collections.Lists2;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.FunctionName;
+import io.crate.types.TypeSignature;
+
+import java.util.Collections;
+import java.util.List;
+
+public final class Signature {
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private FunctionName name;
+        private FunctionInfo.Type kind;
+        private List<TypeSignature> argumentTypes = Collections.emptyList();
+        private TypeSignature returnType;
+        private List<TypeVariableConstraint> typeVariableConstraints = Collections.emptyList();
+        private List<TypeSignature> variableArityGroup = Collections.emptyList();
+        private boolean variableArity = false;
+
+        public Builder name(String name) {
+            return name(new FunctionName(null, name));
+        }
+
+        public Builder name(FunctionName name) {
+            this.name = name;
+            return this;
+        }
+
+        public Builder kind(FunctionInfo.Type kind) {
+            this.kind = kind;
+            return this;
+        }
+
+        public Builder argumentTypes(TypeSignature... argumentTypes) {
+            return argumentTypes(List.of(argumentTypes));
+        }
+
+        public Builder argumentTypes(List<TypeSignature> argumentTypes) {
+            this.argumentTypes = argumentTypes;
+            return this;
+        }
+
+        public Builder returnType(TypeSignature returnType) {
+            this.returnType = returnType;
+            return this;
+        }
+
+        public Builder typeVariableConstraints(TypeVariableConstraint... typeVariableConstraints) {
+            return typeVariableConstraints(List.of(typeVariableConstraints));
+        }
+
+        public Builder typeVariableConstraints(List<TypeVariableConstraint> typeVariableConstraints) {
+            this.typeVariableConstraints = typeVariableConstraints;
+            return this;
+        }
+
+        public Builder variableArityGroup(List<TypeSignature> variableArityGroup) {
+            this.variableArityGroup = variableArityGroup;
+            this.variableArity = !variableArityGroup.isEmpty();
+            return this;
+        }
+
+        public Builder setVariableArity(boolean variableArity) {
+            this.variableArity = variableArity;
+            return this;
+        }
+
+        public Signature build() {
+            return new Signature(
```

Comment:
Can you add some assertions that make sure all required properties are set? (name, kind, argument types..)